File: labs/lab11/Course_Schedule_2.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {number[]}
 */
function findOrder(numCourses, prerequisites) {
  // Step 1: Build the graph and in-degree array
  const graph = new Map();
  const inDegree = Array(numCourses).fill(0);

  for (const [course, pre] of prerequisites) {
    if (!graph.has(pre)) graph.set(pre, []);
    graph.get(pre).push(course);
    inDegree[course]++;
  }

  // Step 2: Initialize the queue with courses that have no prerequisites
  const queue = [];
  for (let i = 0; i < numCourses; i++) {
    if (inDegree[i] === 0) queue.push(i);
  }

  const courseOrder = [];

  // Step 3: Process the courses
  while (queue.length > 0) {
    const current = queue.shift();
    courseOrder.push(current);

    if (graph.has(current)) {
      for (const neighbor of graph.get(current)) {
        inDegree[neighbor]--;
        if (inDegree[neighbor] === 0) {
          queue.push(neighbor);
        }
      }
    }
  }

  // Step 4: Check if all courses are processed
  return courseOrder.length === numCourses ? courseOrder : [];
}

// Example usage
console.log(findOrder(2, [[1, 0]])); // Output: [0, 1]
console.log(
  findOrder(4, [
    [1, 0],
    [2, 0],
    [3, 1],
    [3, 2],
  ])
); // Output: [0, 1, 2, 3] or [0, 2, 1, 3]
console.log(findOrder(1, [])); // Output: [0]

File: labs/lab11/Course_Schedule_2.py
from collections import defaultdict, deque

class Solution:
    def findOrder(self, numCourses, prerequisites):
        #bfs
        # Step 1: Create graph and track prerequisites for each course
        course_graph = defaultdict(list)
        in_degree = [0] * numCourses  # Track the number of prerequisites for each course
        
        # Build the graph and update in-degree counts
        for course, prerequisite in prerequisites:
            course_graph[prerequisite].append(course)
            in_degree[course] += 1
        
        # Step 2: Add courses with no prerequisites to the queue
        queue = deque([course for course in range(numCourses) if in_degree[course] == 0])
        course_order = []  # List to store the final course order
        
        # Step 3: Process courses using BFS
        while queue:
            current_course = queue.popleft()
            course_order.append(current_course)
            
            # Decrease the in-degree of neighboring courses
            for dependent_course in course_graph[current_course]:
                in_degree[dependent_course] -= 1
                if in_degree[dependent_course] == 0:
                    queue.append(dependent_course)
        
        # Step 4: Check if all courses have been processed
        if len(course_order) == numCourses:
            return course_order  # Return valid course order
        return []  # Return an empty list if not all courses can be completed
